src/irr/video/COpenGLSpecializedShader.h

```diff
@@ -52,22 +55,21 @@ class COpenGLSpecializedShader : public core::impl::ResolveAlignment<IGPUSpecial
 
 	private:
 		//! @returns GL name or zero if already compiled once or there were compilation errors.
+		//to del
 		GLuint compile(uint32_t _GLSLversion, const asset::ICPUBuffer* _spirv, const asset::ISpecializedShader::SInfo& _specInfo, const asset::CIntrospectionData* _introspection);
 
 	private:
-		mutable core::smart_refctd_dynamic_array<GLuint> m_GLnames;
+		mutable core::smart_refctd_dynamic_array<GLuint> m_GLnames;//to del
 		GLenum m_GLstage;
-		SProgramBinary m_binary;
+		SProgramBinary m_binary;//to del
 
-		using SMember = asset::impl::SShaderMemoryBlock::SMember;
-		struct SUniform {
-			SUniform(const SMember& _m, GLint _loc, uint8_t* _valptr) : m(_m), location(_loc), value(_valptr) {}
-			SMember m;
-			GLint location;
-			uint8_t* value;
-		};
-		mutable bool m_uniformsSetForTheVeryFirstTime = true;
-		alignas(128) uint8_t m_uniformValues[IGPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE];
+		SInfo m_specInfo;
+		core::smart_refctd_ptr<const asset::ICPUBuffer> m_spirv;
+		std::array<uint64_t, 4> m_spirvHash;
```

Comment:
shouldn't hash be in `COpenGLShader` the unspecialized one?

Reply:
also the SPIR-V?

Reply:
no, why? unspecialized shader has nothing to do with pipelines + unspecialized shader holds spirv OR glsl, while specialized always holds spirv

Reply:
> unspecialized shader holds spirv OR glsl

hash of either is fine.

Are there any mutations that get applied to the GLSL after we create an `IGPUShader` ?

Reply:
well, yea, like everything. #include resolvements, binding reorders, push constants to uniform translation, specialization constants resolvements

Reply:
> binding reorders, push constants to uniform translation, specialization constants resolvement

That's on the GLSL generated from SPIR-V.

But yeah I remember the run-time extension `include`s